### DISCUSSION TITLE
Remote seeding with Supabase CLI

### DISCUSSION THREAD

**Discussion by dilly-sixmo** [supabase/supabase#13832] [Questions] [answered]
Is it possible to use the Supabase CLI to seed my remote database?  I know it can do it locally, with the "supabase db reset" command, which runs your migrations and executes the seed.sql file on your local instance.  However "supabase db push" only pushes migrations to the remote instance, how do I seed it?

I searched the docs and asked the new Supabase AI, and Supabase AI told me that it was possible, and that I just had to use "supabase db seed".  Well that command doesn't exist, so that shows you how much the AI knows!

**Comment by dilly-sixmo** [comment]
Also, how do I pull down migrations?  Is it possible to pull down the migrations saved in the remote "supabase_migrations" schema back down to have it recreate my local migrations folder?  The Supabase AI said I could use "supabase db pull" but once again, that command doesn't exist.

**Reply by silentworks** [reply]
You can pull your initial migration down from your remote project (https://supabase.com/docs/reference/cli/supabase-db-remote-commit) but not the individual migrations as the remote database only records which migration files have been ran already and not the actual change they make (this is standard practice for database schema migrations). The migration files should be kept in your code base repository so you can rerun them if you need to in other environments.

**Comment by silentworks** [comment, marked as answer]
The Supabase CLI doesn't currently support seeding your remote database but you can do this with any other seeding tool that allows you to connect to a Postgres database by using the connection string in the database section of the dashboard https://app.supabase.com/project/_settings/database.

**Reply by rienheuver** [reply]
I believe it does support so nowadays via `supabase db reset --db-url <DB_URL>`. Seeding users crashed for me though, so I had to insert those manually (and all relations to users/profiles therefore as well...)

**Reply by tomelliot** [reply]
@rienheuver I was looking for something like this so this is helpful, thankyou. I just did a quick test and it worked for some data in the public schema, but I still see user data in the `auth.users` table. ~Do you know what will be reset and what won't be reset by this command? Or where I can go for more info (aside from [the docs page](https://supabase.com/docs/reference/cli/supabase-db-reset))?~

I read a bit further. From [the PR](https://github.com/supabase/cli/pull/1105):
>     ignores all supabase managed schemas, including auth, storage, extensions, etc.
>     only drop objects in public schema to preserve default schema privileges for supabase_admin
>     applies all location migrations and seeds remote database

**Comment by imsailoradmin** [comment]
Am I correct to say that we still dont have a proper way to fully reset and reseed remote (staging in my case) whenver we want to? 

I tried the reset command, but I am getting extensions error.

**Reply by MossabDiae** [reply]
reset remote (probably only public schema) `supabase db reset --linked`
push local seed data `supabase db push  --include-seed`

**Reply by hedleysmith** [reply]
This is the answer

**Comment by rasinmuhammed** [comment]
Adding a note for anyone landing here from a search, since the answer has changed twice and the thread is long.

What works today: `supabase db reset --linked` to reset the remote public schema, and `supabase db push --include-seed` to push `seed.sql` with it.

The part that still bites, and which the comment above touches on: reset only clears `public`. Your `auth.users` rows survive, and they are different in every environment. So the moment `seed.sql` inserts anything that references `auth.users(id)`, you are hardcoding UUIDs that are correct in exactly one database. On a preview branch with freshly signed-up test users, the same file either fails the foreign key or silently attaches every row to whoever happens to hold that id.

The approach that has held up for me is to stop hardcoding and read `auth.users` at seed time, then draw the foreign keys from the ids that are actually there. Nothing writes to `auth`, which stays the auth service's job.

I maintain an open source generator that does this, so treat the following as disclosure rather than a recommendation. `pip install "misata[db]"`, then:

```
misata seed "postgresql://postgres:PASS@db.PROJECT.supabase.co:5432/postgres" --dry-run
```

It reads the schema from the database, notices that `profiles.id` points out of `public` at `auth.users`, and treats that table as read-only. Against a local Postgres shaped like a Supabase project (25 auth users, profiles keyed to them, projects and tasks below), the plan it printed was:

```
 #  table     existing  will insert
 1  users            0            0      <- auth.users, read only
 2  profiles         0           25
 3  projects         0        2,500
 4  tasks            0        1,500
```

Note the 25. `profiles.id` is both the primary key and the foreign key, so it is one-to-one with `auth.users` and cannot exceed it. Asking for 2,500 profiles against 25 auth users is not a shortfall to top up later, it is arithmetic that cannot hold, and it should be refused rather than attempted. Drop `--dry-run` and it writes, then re-queries the database to confirm every foreign key resolves.

Two honest limits. I tested this against a local Postgres built to match the shape, not against a hosted Supabase project, so I cannot vouch for the pooled connection on port 6543. Use the direct connection string on 5432 if you hit something odd, and I would like to hear about it if you do. And it does not create auth users, restore storage objects, or write RLS policies. If `auth.users` is empty it refuses and tells you to sign some users up first, because there is nothing valid it could insert.

Even if you never touch any of that: read your auth ids at seed time rather than pasting them into `seed.sql`. That one change is what makes a seed file survive being run against a branch.
